### PR TITLE
Define testWidgetsWithLeakTracking.

### PR DESCRIPTION
**Pull request by polina-c**
contributes to: https://github.com/flutter/devtools/issues/3951

### CODE REVIEW

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (goderbauer) [inline packages/flutter/test/foundation/leak_tracking.dart:12]: Looks like this is only needed for the `Timeout` type below? I believe that type is exported from `package:flutter_test/flutter_test.dart` above somewhere, so you shouldn't need this...

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (goderbauer) [inline packages/flutter/test/foundation/leak_tracking.dart:83]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#only-log-actionable-messages-to-the-console

File: packages/flutter/test/material/tooltip_test.dart
Comment (goderbauer) [inline packages/flutter/test/material/tooltip_test.dart:1024]: nit: remove blank line

File: packages/flutter/test/material/tooltip_test.dart
Comment (goderbauer) [inline packages/flutter/test/material/tooltip_test.dart:1028]: This makes me question that turning on leak tracking by default is actually feasible. This is a fairly common pattern. If people are confronted with a non-actionable leak warnings every time they obtain a state in a test, they'll likely get frustrated with these warnings and stop paying attention to the legitimate ones.

Comment (polina-c) [reply packages/flutter/test/material/tooltip_test.dart:1028]: Good question. 
We can add allowlist for such leaks and suggest the code to copy-paste in error message.

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (polina-c) [reply packages/flutter/test/foundation/leak_tracking.dart:12]: nope. I am getting errors without it.

Comment (goderbauer) [reply packages/flutter/test/foundation/leak_tracking.dart:12]: What's the error? That's something we should fix because users will otherwise not be able to use this without importing this deprecated packages either.

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (polina-c) [reply packages/flutter/test/foundation/leak_tracking.dart]: Made actionable. Peer PR in leak tracker: https://github.com/dart-lang/leak_tracker/pull/48. Please, review.

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (polina-c) [reply packages/flutter/test/foundation/leak_tracking.dart]: Undefined class 'test_package.Timeout'.
Try changing the name to the name of an existing class, or creating a class with the name 'test_package.Timeout'.

Comment (polina-c) [reply packages/flutter/test/foundation/leak_tracking.dart]: fixed. thanks

**Comment by polina-c** [comment]
Prerequisite: https://github.com/flutter/flutter/pull/125225

**Comment by polina-c** [comment]
Added temporal debug prints. Will remove before merge.

File: packages/flutter/test/material/tooltip_test.dart
Comment (dnfield) [reply packages/flutter/test/material/tooltip_test.dart]: We talked about this a bit offline but unfortunately without @goderbauer - this test is a _little_ strange in that it's poking at a disposed object after the object has been disposed.

I think a reasonable API would be to somehow add that object as "intentionally leaked", e.g. `tester.ignoredLeakedObject(tooltipState)`. WDYT @goderbauer ?

Comment (goderbauer) [reply packages/flutter/test/material/tooltip_test.dart]: > this test is a little strange in that it's poking at a disposed object after the object has been disposed.

For my own understanding: If you'd remove the last `expect(tooltipState.ensureTooltipVisible(), false);` from the test, the leak tracker would not complain? Or would holding the disposed `tooltipState` in that variable still constitute a leak? I believe, doing the latter is somewhat common...

> I think a reasonable API would be to somehow add that object as "intentionally leaked",

Again, I think to answer this question we need to know how common this situation is, i.e. how often would test authors get confused about a false leak warning? For that, I would check how often this situation happens in our current test code base...

Comment (polina-c) [reply packages/flutter/test/material/tooltip_test.dart]: I collected some statistics. Out of 93 tests, there is just 1 that holds state after disposal.
Yes, it counts as leak, because an object is expected to be GCed soon after disposal.

Looking forward to discuss details on a meeting.

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (goderbauer) [inline packages/flutter/test/foundation/leak_tracking.dart:10]: typo: garbafge

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (goderbauer) [inline packages/flutter/test/foundation/leak_tracking.dart:38]: Do we still need this with our current decision to not track non-gced leaks? Currently, it doesn't matter whether you mark something as held object or not, right? If so, let's remove this for now to avoid that it is used incorrectly until it actually matters.

Comment (goderbauer) [reply packages/flutter/test/foundation/leak_tracking.dart:38]: Looks like this is also unused in this PR?

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (goderbauer) [inline packages/flutter/test/foundation/leak_tracking.dart:64]: Do we have an issue on file that details why this is? Would be nice to link it from here.

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (goderbauer) [inline packages/flutter/test/foundation/leak_tracking.dart:68]: typo: tracling

File: packages/flutter/test/foundation/leak_tracking_test.dart
Comment (goderbauer) [inline packages/flutter/test/foundation/leak_tracking_test.dart:97]: We avoid checking in commented out code because it just bit-rots. Either remove this test or mark it as skipped.

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (goderbauer) [inline packages/flutter/test/foundation/leak_tracking.dart:62]: The documentation could be improved, see also https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#documentation-dartdocs-javadocs-etc.

The doc comment should give an understanding of WHAT a method does without requiring the reader to chase down a bunch of links (links can give additional context, though, for someone who wants to do a deep dive). 

Also, how we plan to evolve a method (currently the majority of this doc) is not super-relevant to the reader, who is trying to understand how a test with this method works. Ideally, we just summarize that plan in a quick TODO note and link to an issue that contains more information about the next steps.

**Review by goderbauer — APPROVED** [review]
LGTM after comments are addressed.

File: packages/flutter/test/foundation/leak_tracking.dart
Comment (goderbauer) [inline packages/flutter/test/foundation/leak_tracking.dart:14]: Looks like this is also unused now and should have been removed?

**Pull request merged**
